### ISSUE TITLE
DRIVER_IRQL_NOT_LESS_OR_EQUAL or PAGE_FAULT_IN_NONPAGED_AREA bug check on Windows 10 14316 

### ISSUE THREAD

**Issue by tandasat** [tandasat/GuardMon#2]
UtilIsAccessibleAddress() fails to retrieve correct page table entries due to changes in addresses of PXE, PPE etc. It results in accessing an invalid address as a PxE entry and causing a bug check.
### Addresses currently programmed

```
1: kd> !pte
                                           VA 0000000000000000
PXE at FFFFF6FB7DBED000    PPE at FFFFF6FB7DA00000    PDE at FFFFF6FB40000000    PTE at FFFFF68000000000
Unable to get PXE FFFFF6FB7DBED000
```
### Addresses being used in Windows 10 14316

```
1: kd> u nt!MiGetPhysicalAddress l50
nt!MiGetPhysicalAddress:
...
fffff802`fdd71897 48b900409f3e7dfaffff mov rcx,0FFFFFA7D3E9F4000h
...
fffff802`fdd718b2 48b90000803e7dfaffff mov rcx,0FFFFFA7D3E800000h
...
fffff802`fdd718d0 48b9000000007dfaffff mov rcx,0FFFFFA7D00000000h

```
### Error log

```
*******************************************************************************
*                                                                             *
*                        Bugcheck Analysis                                    *
*                                                                             *
*******************************************************************************

Use !analyze -v to get detailed debugging information.

BugCheck D1, {fffff6fb7dbed000, ff, 85, fffff8064d25420c}

*** ERROR: Module load completed but symbols could not be loaded for mssmbios.sys
Probably caused by : GuardMon.sys ( GuardMon!UtilIsAccessibleAddress+4c )

Followup:     MachineOwner
---------

1: kd> !analyze -v
*******************************************************************************
*                                                                             *
*                        Bugcheck Analysis                                    *
*                                                                             *
*******************************************************************************

DRIVER_IRQL_NOT_LESS_OR_EQUAL (d1)
An attempt was made to access a pageable (or completely invalid) address at an
interrupt request level (IRQL) that is too high.  This is usually
caused by drivers using improper addresses.
If kernel debugger is available get stack backtrace.
Arguments:
Arg1: fffff6fb7dbed000, memory referenced
Arg2: 00000000000000ff, IRQL
Arg3: 0000000000000085, value 0 = read operation, 1 = write operation
Arg4: fffff8064d25420c, address which referenced memory

Debugging Details:
------------------


DUMP_CLASS: 1

DUMP_QUALIFIER: 401

BUILD_VERSION_STRING:  14316.1000.amd64fre.rs1_release.160402-2217

DUMP_TYPE:  1

BUGCHECK_P1: fffff6fb7dbed000

BUGCHECK_P2: ff

BUGCHECK_P3: 85

BUGCHECK_P4: fffff8064d25420c

WRITE_ADDRESS:  fffff6fb7dbed000 

CURRENT_IRQL:  2

FAULTING_IP: 
GuardMon!UtilIsAccessibleAddress+4c [c:\users\user\desktop\guardmon\hyperplatform\hyperplatform\util.cpp @ 448]
fffff806`4d25420c 488b00          mov     rax,qword ptr [rax]

CPU_COUNT: 4

CPU_MHZ: 8fc

CPU_VENDOR:  GenuineIntel

CPU_FAMILY: 6

CPU_MODEL: 45

CPU_STEPPING: 1

CPU_MICROCODE: 6,45,1,0 (F,M,S,R)  SIG: 1D'00000000 (cache) 1D'00000000 (init)

DEFAULT_BUCKET_ID:  WIN8_DRIVER_FAULT

BUGCHECK_STR:  AV

PROCESS_NAME:  System

ANALYSIS_SESSION_HOST:  DESKTOP-LQSEFPE

ANALYSIS_SESSION_TIME:  04-16-2016 17:00:41.0382

ANALYSIS_VERSION: 10.0.10586.567 amd64fre

TRAP_FRAME:  ffffb800ab805b10 -- (.trap 0xffffb800ab805b10)
NOTE: The trap frame does not contain all registers.
Some register values may be zeroed or incorrect.
rax=fffff6fb7dbed000 rbx=0000000000000000 rcx=fffff6fb7da00008
rdx=ffffb800ab805c80 rsi=0000000000000000 rdi=0000000000000000
rip=fffff8064d25420c rsp=ffffb800ab805ca0 rbp=ffffa581753b3b00
 r8=ffffb800ab805c78  r9=ffffb800ab805bd0 r10=0000000000000000
r11=ffffb800ab8056d0 r12=0000000000000000 r13=0000000000000000
r14=0000000000000000 r15=0000000000000000
iopl=0         nv up di ng nz na po nc
GuardMon!UtilIsAccessibleAddress+0x4c:
fffff806`4d25420c 488b00          mov     rax,qword ptr [rax] ds:fffff6fb`7dbed000=????????????????
Resetting default scope

BAD_STACK_POINTER:  ffffb800ab8059c8

LAST_CONTROL_TRANSFER:  from fffff802fdde7629 to fffff802fdddc4b0

STACK_TEXT:  
ffffb800`ab8059c8 fffff802`fdde7629 : 00000000`0000000a fffff6fb`7dbed000 00000000`000000ff 00000000`00000085 : nt!KeBugCheckEx
ffffb800`ab8059d0 fffff802`fdde5c07 : 00000000`00000001 fffff806`4d254620 00000000`00000001 00000000`00000000 : nt!KiBugCheckDispatch+0x69
ffffb800`ab805b10 fffff806`4d25420c : 00000000`5058768b fffff806`4d25174a 00000000`00000000 ffffb800`ab805fe0 : nt!KiPageFault+0x247
ffffb800`ab805ca0 fffff806`4d25ac64 : 00000000`5058768b 00000000`00000000 0c7f2cae`7e56dce0 fffff806`4d251725 : GuardMon!UtilIsAccessibleAddress+0x4c [c:\users\user\desktop\guardmon\hyperplatform\hyperplatform\util.cpp @ 448]
ffffb800`ab805d10 fffff806`4d25a887 : 00000000`505875c3 fffff806`4d253ccc ffffa581`753b3b00 00000000`002245ce : GuardMon!GMonpIsPgContext+0x54 [c:\users\user\desktop\guardmon\guardmon\guard_mon.cpp @ 279]
ffffb800`ab805d60 fffff806`4d25aa89 : ffffb800`ab805f70 00000000`00000002 00000000`00000002 fffff806`4d25a852 : GuardMon!GMonIsPgExcutionContext+0x17 [c:\users\user\desktop\guardmon\guardmon\guard_mon.cpp @ 249]
ffffb800`ab805d90 fffff806`4d256eb3 : ffffb800`ab805f70 fffff806`4d25b870 fffff806`4d25b850 ffffb800`ace12946 : GuardMon!GMonRemoveNoCr0ModificationFlag+0x29 [c:\users\user\desktop\guardmon\guardmon\guard_mon.cpp @ 160]
ffffb800`ab805df0 fffff806`4d2573bd : ffffb800`ab805f18 ffffb800`9cc95000 fffff806`4d253cc0 fffff806`4d25b740 : GuardMon!VmmpHandleRdtsc+0xc3 [c:\users\user\desktop\guardmon\hyperplatform\hyperplatform\vmm.cpp @ 437]
ffffb800`ab805e60 fffff806`4d254f7b : ffffb800`ab805f18 00000000`00000000 00000000`00000000 00000000`00000000 : GuardMon!VmmpHandleVmExit+0x1bd [c:\users\user\desktop\guardmon\hyperplatform\hyperplatform\vmm.cpp @ 240]
ffffb800`ab805ee0 fffff806`4d251345 : ffffb800`ab805f70 00000000`00000000 00000000`00000000 00000000`00000000 : GuardMon!VmmVmExitHandler+0xcb [c:\users\user\desktop\guardmon\hyperplatform\hyperplatform\vmm.cpp @ 186]
ffffb800`ab805f50 ffffb800`ab805f70 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`fffffff8 : GuardMon!AsmVmmEntryPoint+0x25 [C:\Users\user\Desktop\GuardMon\HyperPlatform\HyperPlatform\Arch\x64\x64.asm @ 148]
ffffb800`ab805f58 00000000`00000000 : 00000000`00000000 00000000`00000000 00000000`fffffff8 00000000`00000040 : 0xffffb800`ab805f70


STACK_COMMAND:  kb

THREAD_SHA1_HASH_MOD_FUNC:  bc6b408429ca6159ee390dd09f2c0be6a334e7a8

THREAD_SHA1_HASH_MOD_FUNC_OFFSET:  8febc4ec83ec099c8a2d2aedf2f7034544466616

THREAD_SHA1_HASH_MOD:  ab245e343572bc9a476077682ac489aa36af2a34

FOLLOWUP_IP: 
GuardMon!UtilIsAccessibleAddress+4c [c:\users\user\desktop\guardmon\hyperplatform\hyperplatform\util.cpp @ 448]
fffff806`4d25420c 488b00          mov     rax,qword ptr [rax]

FAULT_INSTR_CODE:  48008b48

FAULTING_SOURCE_LINE:  c:\users\user\desktop\guardmon\hyperplatform\hyperplatform\util.cpp

FAULTING_SOURCE_FILE:  c:\users\user\desktop\guardmon\hyperplatform\hyperplatform\util.cpp

FAULTING_SOURCE_LINE_NUMBER:  448

FAULTING_SOURCE_CODE:  
   444: 
   445: #if defined(_AMD64_)
   446:   const auto pxe = UtilpAddressToPxe(address);
   447:   const auto ppe = UtilpAddressToPpe(address);
>  448:   if (!pxe || !pxe->valid || !ppe || !ppe->valid) {
   449:     return false;
   450:   }
   451: #endif
   452: 
   453:   const auto is_x86_pae = UtilIsX86Pae();


SYMBOL_STACK_INDEX:  3

SYMBOL_NAME:  GuardMon!UtilIsAccessibleAddress+4c

FOLLOWUP_NAME:  MachineOwner

MODULE_NAME: GuardMon

IMAGE_NAME:  GuardMon.sys

DEBUG_FLR_IMAGE_TIMESTAMP:  5712c871

BUCKET_ID_FUNC_OFFSET:  4c

FAILURE_BUCKET_ID:  AV_STACKPTR_ERROR_GuardMon!UtilIsAccessibleAddress

BUCKET_ID:  AV_STACKPTR_ERROR_GuardMon!UtilIsAccessibleAddress

PRIMARY_PROBLEM_CLASS:  AV_STACKPTR_ERROR_GuardMon!UtilIsAccessibleAddress

TARGET_TIME:  2016-04-16T23:23:16.000Z

OSBUILD:  14316

OSSERVICEPACK:  0

SERVICEPACK_NUMBER: 0

OS_REVISION: 0

SUITE_MASK:  272

PRODUCT_TYPE:  1

OSPLATFORM_TYPE:  x64

OSNAME:  Windows 10

OSEDITION:  Windows 10 WinNt TerminalServer SingleUserTS

OS_LOCALE:  

USER_LCID:  0

OSBUILD_TIMESTAMP:  2016-04-03 00:42:04

BUILDDATESTAMP_STR:  160402-2217

BUILDLAB_STR:  rs1_release

BUILDOSVER_STR:  10.0.14316.1000.amd64fre.rs1_release.160402-2217

ANALYSIS_SESSION_ELAPSED_TIME: 1f4c

ANALYSIS_SOURCE:  KM

FAILURE_ID_HASH_STRING:  km:av_stackptr_error_guardmon!utilisaccessibleaddress

FAILURE_ID_HASH:  {c3adc55d-0eb2-f557-a195-a9273d61984d}

Followup:     MachineOwner
---------

```

**Comment by tandasat** [comment]
This is another instance by MemoryMon using the same piece of code not supporting Windows 10 14316 proper.

```
*******************************************************************************
*                                                                             *
*                        Bugcheck Analysis                                    *
*                                                                             *
*******************************************************************************

PAGE_FAULT_IN_NONPAGED_AREA (50)
Invalid system memory was referenced.  This cannot be protected by try-except.
Typically the address is just plain bad or it is pointing at freed memory.
Arguments:
Arg1: fffff6fb7dbedf80, memory referenced.
Arg2: 0000000000000000, value 0 = read operation, 1 = write operation.
Arg3: fffff80cb903265d, If non-zero, the instruction address which referenced the bad memory
    address.
Arg4: 0000000000000002, (reserved)

Debugging Details:
------------------


DUMP_CLASS: 1

DUMP_QUALIFIER: 401

BUILD_VERSION_STRING:  14316.1000.amd64fre.rs1_release.160402-2217

DUMP_TYPE:  1

BUGCHECK_P1: fffff6fb7dbedf80

BUGCHECK_P2: 0

BUGCHECK_P3: fffff80cb903265d

BUGCHECK_P4: 2

READ_ADDRESS:  fffff6fb7dbedf80 

FAULTING_IP: 
MemoryMon!UtilIsNonPageableAddress+21 [c:\users\user\desktop\hp\memorymon\hyperplatform\hyperplatform\util.cpp @ 502]
fffff80c`b903265d 8a04c2          mov     al,byte ptr [rdx+rax*8]

MM_INTERNAL_CODE:  2

IMAGE_NAME:  MemoryMon.sys

DEBUG_FLR_IMAGE_TIMESTAMP:  5715c973

MODULE_NAME: MemoryMon

FAULTING_MODULE: fffff80cb9030000 MemoryMon

CPU_COUNT: 4

CPU_MHZ: 8fc

CPU_VENDOR:  GenuineIntel

CPU_FAMILY: 6

CPU_MODEL: 45

CPU_STEPPING: 1

CPU_MICROCODE: 6,45,1,0 (F,M,S,R)  SIG: 1D'00000000 (cache) 1D'00000000 (init)

DEFAULT_BUCKET_ID:  WIN8_DRIVER_FAULT

BUGCHECK_STR:  AV

PROCESS_NAME:  System

CURRENT_IRQL:  0

ANALYSIS_SESSION_HOST:  DESKTOP-LQSEFPE

ANALYSIS_SESSION_TIME:  04-19-2016 07:29:18.0234

ANALYSIS_VERSION: 10.0.10586.567 amd64fre

TRAP_FRAME:  ffffdb81acda8230 -- (.trap 0xffffdb81acda8230)
NOTE: The trap frame does not contain all registers.
Some register values may be zeroed or incorrect.
rax=00000000000001f0 rbx=0000000000000000 rcx=fffff80cb39ea000
rdx=fffff6fb7dbed000 rsi=0000000000000000 rdi=0000000000000000
rip=fffff80cb903265d rsp=ffffdb81acda83c8 rbp=fffffa8000000000
 r8=fffff80cb39ea000  r9=fffffa8000000000 r10=ffffdb81acda83d8
r11=fffff80cb9034901 r12=0000000000000000 r13=0000000000000000
r14=0000000000000000 r15=0000000000000000
iopl=0         nv up ei pl nz na po nc
MemoryMon!UtilIsNonPageableAddress+0x21:
fffff80c`b903265d 8a04c2          mov     al,byte ptr [rdx+rax*8] ds:fffff6fb`7dbedf80=??
Resetting default scope

LAST_CONTROL_TRANSFER:  from fffff80285806bd8 to fffff802857d34b0

STACK_TEXT:  
ffffdb81`acda7f48 fffff802`85806bd8 : 00000000`00000050 fffff6fb`7dbedf80 00000000`00000000 ffffdb81`acda8230 : nt!KeBugCheckEx
ffffdb81`acda7f50 fffff802`856f3fab : 00000000`00000000 00000600`00000000 ffffdb81`acda8230 fffffd80`00001c00 : nt! ?? ::FNODOBFM::`string'+0x24588
ffffdb81`acda8040 fffff802`857dcafc : 00000000`0000000c 00000000`000002a0 00000000`00000001 00000000`00000000 : nt!MmAccessFault+0x8db
ffffdb81`acda8230 fffff80c`b903265d : fffff80c`b903fc1f ffffcb80`449781d0 fffff80c`b39a0000 00000000`001ca000 : nt!KiPageFault+0x13c
ffffdb81`acda83c8 fffff80c`b903fc1f : ffffcb80`449781d0 fffff80c`b39a0000 00000000`001ca000 00000000`00000200 : MemoryMon!UtilIsNonPageableAddress+0x21 [c:\users\user\desktop\hp\memorymon\hyperplatform\hyperplatform\util.cpp @ 502]
ffffdb81`acda83d0 fffff80c`b903e490 : ffffcb80`4502d8c0 ffffcb80`4502d8c0 ffffcb80`4e9437f0 ffffcb80`4d71c000 : MemoryMon!MmoneptInitialization+0x107 [c:\users\user\desktop\hp\memorymon\memorymon\memorymon_ept.cpp @ 122]
ffffdb81`acda8570 fffff80c`b903edee : 00000000`00000028 ffffdb81`acda8760 ffffcb80`454cb6a0 00000000`00000000 : MemoryMon!EptInitialization+0x1e8 [c:\users\user\desktop\hp\memorymon\hyperplatform\hyperplatform\ept.cpp @ 275]
ffffdb81`acda85c0 fffff80c`b903e033 : ffffcb80`4de2d660 ffffdb81`a4797bc0 00000000`00000000 ffffffff`80001e44 : MemoryMon!VmpInitializeVm+0x62 [c:\users\user\desktop\hp\memorymon\hyperplatform\hyperplatform\vm.cpp @ 306]
ffffdb81`acda8630 ffffcb80`4de2d660 : ffffdb81`a4797bc0 00000000`00000000 ffffffff`80001e44 ffffcb80`4de2d660 : MemoryMon!AsmInitializeVm+0x33 [C:\Users\user\Desktop\HP\MemoryMon\HyperPlatform\HyperPlatform\Arch\x64\x64.asm @ 112]
ffffdb81`acda8638 ffffdb81`a4797bc0 : 00000000`00000000 ffffffff`80001e44 ffffcb80`4de2d660 fffff80c`b903f9ec : 0xffffcb80`4de2d660
ffffdb81`acda8640 00000000`00000000 : ffffffff`80001e44 ffffcb80`4de2d660 fffff80c`b903f9ec 00000000`00000200 : 0xffffdb81`a4797bc0


STACK_COMMAND:  kb

THREAD_SHA1_HASH_MOD_FUNC:  a147cbd486fd8a863d9715481245383cee99f2c0

THREAD_SHA1_HASH_MOD_FUNC_OFFSET:  dbf3d2ce61190529d5116c84d082c4431094e5d0

THREAD_SHA1_HASH_MOD:  5da0c6b0fec031fb46756ffef57dc1a44821663f

FOLLOWUP_IP: 
MemoryMon!UtilIsNonPageableAddress+21 [c:\users\user\desktop\hp\memorymon\hyperplatform\hyperplatform\util.cpp @ 502]
fffff80c`b903265d 8a04c2          mov     al,byte ptr [rdx+rax*8]

FAULT_INSTR_CODE:  41c2048a

FAULTING_SOURCE_LINE:  c:\users\user\desktop\hp\memorymon\hyperplatform\hyperplatform\util.cpp

FAULTING_SOURCE_FILE:  c:\users\user\desktop\hp\memorymon\hyperplatform\hyperplatform\util.cpp

FAULTING_SOURCE_LINE_NUMBER:  502

FAULTING_SOURCE_CODE:  
   498: 
   499: #if defined(_AMD64_)
   500:   const auto pxe = UtilpAddressToPxe(address);
   501:   const auto ppe = UtilpAddressToPpe(address);
>  502:   if (!pxe->valid || !ppe->valid) {
   503:     return false;
   504:   }
   505: #endif
   506: 
   507:   const auto is_x86_pae = UtilIsX86Pae();


SYMBOL_STACK_INDEX:  4

SYMBOL_NAME:  MemoryMon!UtilIsNonPageableAddress+21

FOLLOWUP_NAME:  MachineOwner

BUCKET_ID_FUNC_OFFSET:  21

FAILURE_BUCKET_ID:  AV_R_INVALID_MemoryMon!UtilIsNonPageableAddress

BUCKET_ID:  AV_R_INVALID_MemoryMon!UtilIsNonPageableAddress

PRIMARY_PROBLEM_CLASS:  AV_R_INVALID_MemoryMon!UtilIsNonPageableAddress

TARGET_TIME:  2016-04-19T06:01:34.000Z

OSBUILD:  14316

OSSERVICEPACK:  0

SERVICEPACK_NUMBER: 0

OS_REVISION: 0

SUITE_MASK:  272

PRODUCT_TYPE:  1

OSPLATFORM_TYPE:  x64

OSNAME:  Windows 10

OSEDITION:  Windows 10 WinNt TerminalServer SingleUserTS

OS_LOCALE:  

USER_LCID:  0

OSBUILD_TIMESTAMP:  2016-04-03 00:42:04

BUILDDATESTAMP_STR:  160402-2217

BUILDLAB_STR:  rs1_release

BUILDOSVER_STR:  10.0.14316.1000.amd64fre.rs1_release.160402-2217

ANALYSIS_SESSION_ELAPSED_TIME: cc6

ANALYSIS_SOURCE:  KM

FAILURE_ID_HASH_STRING:  km:av_r_invalid_memorymon!utilisnonpageableaddress

FAILURE_ID_HASH:  {ddb7386a-f405-9c26-4526-7e8546ee6d31}

Followup:     MachineOwner
---------

```

**Comment by tandasat** [comment]
fixed at 079e41cd8ddd34f1e09f12d022b8d90fd94263b0
